Add Cuda Vector Types (#7949)
numba/numba: docs/source/cuda/types.rst

> Types on CUDA target

Reviewer: ```suggestion
CUDA-Specific Types
```

====================

Note: This page is about types specific to CUDA targets. Most types on CPU targets are also

> available in CUDA target, for information on these types, see :ref:`numba-types`.

Reviewer: There is a list of Python / NumPy types supported in the CUDA target at https://numba--7949.org.readthedocs.build/en/7949/cuda/cudapysupported.html#built-in-types - could you update the reference to point there?

Vector Types
~~~~~~~~~~~~

`CUDA Vector Types <https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#built-in-vector-types>`_

> are available in the kernel. There are 2 important distinctions from vector types in numba CUDA:

Reviewer: ```suggestion
are usable in kernels. There are two important distinctions from vector types in CUDA C/C++:
```

> First, the recommended names for vector types in Numba CUDA is formatted as ``base_typexN``,

Reviewer: ```suggestion
First, the recommended names for vector types in Numba CUDA is formatted as ``<base_type>x<N>``,
```

> where `base_type` is the base type of the vector, and `N` is the number of elements in the vector.

Reviewer: Sphinx formatting for code instead of emphasis:

```suggestion
where ``base_type`` is the base type of the vector, and ``N`` is the number of elements in the vector.
```

> Examples include ``int64x3``, ``uint16x4``, ``float32x4``, etc. This is consistent in type names with
> the pydata community where bit widths are explicit. For new Numba CUDA kernels, this is the recommended
> way to write vector types.

Reviewer: ```suggestion
Examples include ``int64x3``, ``uint16x4``, ``float32x4``, etc. For new Numba CUDA kernels,
this is the recommended way to instantiate vector types.
```

> For convenience, users who prefer to adapt existing kernels from native CUDA to Numba CUDA may use
> aliases consistent with its native namings. For example, ``float3`` aliases ``float32x3``,
> ``long3`` aliases ``int32x3`` or ``int64x3``, depending on platform specification, etc. 

Reviewer: ```suggestion
For convenience, users adapting existing kernels from CUDA C/C++ to Python may use
aliases consistent with the C/C++ namings. For example, ``float3`` aliases ``float32x3``,
``long3`` aliases ``int32x3`` or ``int64x3`` (depending on the platform), etc. 
```

> Second, unlike native CUDA where factory functions are used, vector types are constructed directly

Reviewer: ```suggestion
Second, unlike CUDA C/C++ where factory functions are used, vector types are constructed directly
```

with their constructor. For example, to construct a ``float32x3``:

.. code-block:: python3

    from numba.cuda import float32x3

    # In kernel

>     f3 = float32x3(float32(0.0), float32(-1.0), float32(1.0))

Reviewer: Are the casts necessary? What happens if the user does `float32x3(0.0, -1.0, 1.0)`?

Author: It turns out they are not necessary, since I only defined `float32` as the argument, I believe the floating point literals were cast to `float32`.

Additionally, vector types can be constructed from a combination of vector and
primitive types, as long as the total number of components matches the result
vector type. For example, all of the following constructions are valid:

.. code-block:: python3

    zero = uint32(0)
    u2 = uint32x2(1, 2)
    # Construct a 3-component vector with primitive type and a 2-component vector
    u3 = uint32x3(zero, u2)
    # Construct a 4-component vector with 2 2-component vectors
    u4 = uint32x4(u2, u2)

The 1st, 2nd, 3rd and 4th component of the vector type can be accessed through fields 

> ``x``, ``y``, ``z``, and ``w`` respectively. The components are immutable after

Author: Great idea
